core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/OnDemandRepairSchedulerImpl.java

```diff
@@ -0,0 +1,195 @@
+ * limitations under the License.
+ */
+package com.ericsson.bss.cassandra.ecchronos.core.repair;
+
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.ericsson.bss.cassandra.ecchronos.core.JmxProxyFactory;
+import com.ericsson.bss.cassandra.ecchronos.core.metrics.TableRepairMetrics;
+import com.ericsson.bss.cassandra.ecchronos.core.repair.state.ReplicationState;
+import com.ericsson.bss.cassandra.ecchronos.core.scheduling.ScheduleManager;
+import com.ericsson.bss.cassandra.ecchronos.core.scheduling.ScheduledJob;
+import com.ericsson.bss.cassandra.ecchronos.core.utils.TableReference;
+
+/**
+ * A factory creating {@link OnDemandRepairJob}'s for tables.
+ */
+public class OnDemandRepairSchedulerImpl implements OnDemandRepairScheduler, Closeable
+{
+    private static final Logger LOG = LoggerFactory.getLogger(OnDemandRepairSchedulerImpl.class);
+
+    private final Map<TableReference, OnDemandRepairJob> myScheduledJobs = new HashMap<>();
+    private final Object myLock = new Object();
+
+    private final ExecutorService myExecutor;
+
+    private final JmxProxyFactory myJmxProxyFactory;
+    private final TableRepairMetrics myTableRepairMetrics;
+    private final ScheduleManager myScheduleManager;
+    private final ReplicationState myReplicationState;
+    private final RepairLockType myRepairLockType;
+
+    private OnDemandRepairSchedulerImpl(Builder builder)
+    {
+        myExecutor = Executors.newSingleThreadScheduledExecutor();
+        myJmxProxyFactory = builder.myJmxProxyFactory;
+        myTableRepairMetrics = builder.myTableRepairMetrics;
+        myScheduleManager = builder.myScheduleManager;
+        myReplicationState = builder.myReplicationState;
+        myRepairLockType = builder.repairLockType;
+    }
+
+    @Override
+    public void close()
+    {
+        myExecutor.shutdown();
+        try
+        {
+            if (!myExecutor.awaitTermination(10, TimeUnit.SECONDS))
+            {
+                LOG.warn("Waited 10 seconds for executor to shutdown, still not shut down");
+            }
+        }
+        catch (InterruptedException e)
+        {
+            LOG.error("Interrupted while waiting for executor to shutdown", e);
+            Thread.currentThread().interrupt();
+        }
+
+        synchronized (myLock)
+        {
+            for (TableReference tableReference : myScheduledJobs.keySet())
+            {
+                ScheduledJob job = myScheduledJobs.get(tableReference);
+
+                descheduleTable(job);
+            }
+
+            myScheduledJobs.clear();
+        }
+    }
+
+    @Override
+    public void scheduleJob(TableReference tableReference)
```

Comment:
Should we sanity check that the table exists?

Reply:
Done